### PR TITLE
Add support for case expressions. Add failure tests.

### PR DESCRIPTION
**Pull request by adam-tg**
I implemented the case expression using a series of if statements and gotos (thanks to Finn for the goto idea), which I think is the most readable of the various ways it could have been implemented. The other was handled by setting the default value of the result variable when the other arm is present, and adding a panic if no cases match otherwise.

I implemented failure testing for expected panics, but the logic is still rather rudimentary. Let me know if you want me to fix it up.

### CODE REVIEW

File: test/pgo/IntegrationTestingUtils.java [hunk 1]
```diff
@@ -26,6 +47,97 @@ public static void expungeFile(File file) {
...
+
+			// display the compiled code for inspection
+			Files.lines(compiledOutputPath).forEach(line -> {
+				System.out.println("source: "+line);
+			});
+
+			System.out.println("Got here ok");
+
+			// try to run the compiled Go code, check that it prints the right thing
+			if (!panic) {
```
Comment (fhackett) [inline test/pgo/IntegrationTestingUtils.java]: Thought: for a more meaningful parameterization of this function, we could replace the `panic` argument with a `Consumer<Path>` that you might call like `runner.accept(compiledOutputPath);`.

Example callsite for this function with that change:
```
IntegrationTestingUtils.testCompileExpression(result, vars, compiledOutputPath ->
        IntegrationTestingUtils.testRunGoCode(compiledOutputPath, expected);
```

File: test/pgo/IntegrationTestingUtils.java [hunk 2]
```diff
@@ -26,6 +47,97 @@ public static void expungeFile(File file) {
...
+					generatedConfigPath.toString(),
+					generatedCodePath.toString(),
+			});
+
+			// display the compiled code for inspection
+			Files.lines(compiledOutputPath).forEach(line -> {
+				System.out.println("source: "+line);
+			});
+
+			System.out.println("Got here ok");
```
[the remark this answers is not in the document]
Comment (adam-tg) [reply test/pgo/IntegrationTestingUtils.java]: Whoops, that was a bit of debug code I forgot to remove... I'll quickly delete that.

File: test/pgo/IntegrationTestingUtils.java (hunk 1 above)
Comment (adam-tg) [reply test/pgo/IntegrationTestingUtils.java]: Good idea, that will make it very easy to add new testing functionality.

Comment (adam-tg) [reply test/pgo/IntegrationTestingUtils.java]: Cleaned up by using this interface definition in IntegrationTestingUtils:
```
public interface TestRunner<T> {
	public void run(T t) throws IOException;
}
```

File: src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java [hunk 3]
```diff
@@ -172,7 +172,31 @@ public Expression visit(PGoTLABool pGoTLABool) throws RuntimeException {
...
+		VariableName result = null;
+		if (pGoTLACase.getOther() != null) {
+			result = builder.varDecl("result", pGoTLACase.getOther().accept(this));
+		} else {
+			UID uid = pGoTLACase.getArms().get(0).getResult().getUID();
+			result = builder.varDecl("result", typeMap.get(uid).accept(new PGoTypeGoTypeConversionVisitor()));
+		}
+		LabelName matched = builder.newLabel("matched");
+
+		for (PGoTLACaseArm caseArm : pGoTLACase.getArms()) {
```
Comment (mrordinaire) [inline src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: Does `pGoTLACase.getArms()` return the `OTHER` arm?

```pluscal
CASE x -> y
  [] z -> p
  [] OTHER -> other
```

Comment (adam-tg) [reply src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: It does not (the types wouldn't match). The `OTHER` arm is accessed only by `pGoTLACase.getOther()`.

Comment (mrordinaire) [reply src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: It looks like you're missing codegen for the other branch when it is present.

Comment (adam-tg) [reply src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: The `OTHER` arm is set to the default value of the result variable when present in the first `if`. It's confusing but seemed better than adding a check after the `if`s (which would have either required support for `else if` or a temporary boolean). Now that I'm using `goto`s I could move it below the `if`s but before the label if that would be more clear. What are your thoughts?

Comment (mrordinaire) [reply src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: Got it! I missed that part in the above code. I prefer the translated code to have similar logical structure to the TLA+ code. Could you please put code generation for the `OTHER` arm at the end?

Comment (adam-tg) [reply src/pgo/trans/intermediate/TLAExpressionCodeGenVisitor.java]: Ah, good to know. I just fixed it!

**Pull request merged**